player/hot_cache.go

```diff
@@ -45,92 +46,109 @@ func NewHotCache(origin store.BlobStore, maxSizeInBytes int64) *HotCache {
 
-		h.cache.Set(hash, sized(encoded.Bytes()), longTTL)
+		return &sd, nil
+	})
 
-		return sdBlob, nil
+	if err != nil || blob == nil {
+		return nil, err
 	}
+
+	return blob.(*stream.SDBlob), nil
 }
 
 // GetChunk gets a decrypted stream chunk. If chunk is not cached, it is fetched from origin
 // and decrypted.
 func (h *HotCache) GetChunk(hash string, key, iv []byte) (ReadableChunk, error) {
-	chunk, err, _ := h.sf.Do(hash, h.chunkGetter(hash, key, iv))
-	if err != nil {
-		return nil, err
+	item := h.cache.Get(hash)
+	if item != nil {
+		metrics.HotCacheRequestCount.WithLabelValues("chunk", "hit").Inc()
+		return ReadableChunk(item.Value().(sizedSlice)[:]), nil
 	}
-	return chunk.(ReadableChunk), nil
+
+	metrics.HotCacheRequestCount.WithLabelValues("chunk", "miss").Inc()
+	return h.getChunkFromOrigin(hash, key, iv)
 }
 
-// chunkGetter actually gets the chunk
-func (h *HotCache) chunkGetter(hash string, key, iv []byte) func() (interface{}, error) {
-	return func() (interface{}, error) {
-		item := h.cache.Get(hash)
-		if item != nil {
-			metrics.HotCacheRequestCount.WithLabelValues("chunk", "hit").Inc()
-			return ReadableChunk(item.Value().(sized)), nil
-		}
+// clearChunkFromCache will remove the chunk from the hotcache and from its origin.
+func (h *HotCache) clearChunkFromCache(hash string) error {
+	h.cache.Delete(hash)
+	err := h.origin.Delete(hash)
+	if !errors.Is(err, shared.ErrNotImplemented) {
+		return err
+	}
+	return nil
+}
 
-		metrics.HotCacheRequestCount.WithLabelValues("chunk", "miss").Inc()
-		blob, err := h.origin.Get(hash)
+// getChunkFromOrigin gets the chunk from the origin, decrypts it, caches it, and returns it
+func (h *HotCache) getChunkFromOrigin(hash string, key, iv []byte) (ReadableChunk, error) {
+	chunk, err, _ := h.sf.Do(hash, func() (interface{}, error) {
+		blob, trace, err := h.origin.Get(hash)
 		if err != nil {
 			return nil, err
 		}
-
+		Logger.Debugf("trace for %s:\n%s", hash, trace.String())
 		metrics.InBytes.Add(float64(len(blob)))
 
 		chunk, err := stream.DecryptBlob(blob, key, iv)
 		if err != nil {
 			return nil, err
 		}
 
-		h.cache.Set(hash, sized(chunk), longTTL)
+		h.cache.Set(hash, sizedSlice(chunk), longTTL)
 
 		return ReadableChunk(chunk), nil
+	})
+
+	if err != nil || chunk == nil {
+		return nil, err
 	}
+
+	return chunk.(ReadableChunk)[:], nil
 }
 
 func (h *HotCache) IsCached(hash string) bool {
 	return h.cache.Get(hash) != nil
 }
 
-type sized []byte
+type sizedSlice []byte
+
+func (s sizedSlice) Size() int64 { return int64(len(s)) }
+
+type sizedSD struct {
+	sd *stream.SDBlob
+}
 
-func (s sized) Size() int64 { return int64(len(s)) }
+func (s sizedSD) Size() int64 {
+	total := int64(unsafe.Sizeof(s)) + int64(unsafe.Sizeof(&(s.sd)))
+	for _, bi := range s.sd.BlobInfos {
+		total += int64(unsafe.Sizeof(bi)) + int64(len(bi.BlobHash)+len(bi.IV))
+	}
+	return total + int64(len(s.sd.StreamName)+len(s.sd.StreamType)+len(s.sd.Key)+len(s.sd.SuggestedFileName)+len(s.sd.StreamHash))
+}
```

Comment:
hey @lyoshenka what is this magic? Andrey and I are looking at this but aren't too sure what it's doing (tough we know it works)

Reply:
its the actual size in memory of the stream.SDBlob struct

Reply:
it adds up the size of each field in the struct

i don't think its magic, but it is uncommon. if there's a better way to get the actual size in RAM of a struct, let me know

Reply:
ccache uses Size() to track how much space is used and to know when it needs to evict something

Reply:
Gotcha thank you.